README.md
# Usage
1. Clone `git clone git@bitbucket.org:conversantltd/chump.git && cd chump`
2. Build `mvn clean install`
3. Configure `chump.properties`
3. Run `java -jar target/chump-1.0.0.jar`

> **Important**
> Documentation below is deprecated - see [swagger docs](http://localhost:9090/docs)

# REST - v2
## Create a number (optional - filter.exclude.did) (see swagger docs)
    curl -XPOST -H "Content-type: application/json" http://localhost:9090/chump/v2/numbers -d \
    '{
        "number": "6494297021",
        "domain": "conversant.co.nz",
        "priceListVersionId": 1000000,
        "countryId": "147",
        "countryCode": "64",
        "areaCode": "9429",
        "areaCodeDescription": "Auckland - Red Beach Test",
        "freeMinutes": "10",
        "perMinuteCharge": "0",
        "businessPartnerId": 1000076,
        "setupCost": "1",
        "monthlyCharge": "2",
        "currencyId": 121,
        "orgId": 1000001,
        "name": "DID/DDI"
    }'

    curl -XPOST -H "Content-type: application/json" http://localhost:9090/chump/v2/numbers -d \
    '{
        "numbers": ["6494297022", "6494297023"],
        "domain": "conversant.co.nz",
        "priceListVersionId": 1000000,
        "countryId": "147",
        "countryCode": "64",
        "areaCode": "9429",
        "areaCodeDescription": "Auckland - Red Beach Test",
        "freeMinutes": "10",
        "perMinuteCharge": "0",
        "businessPartnerId": 1000076,
        "setupCost": "1",
        "monthlyCharge": "2",
        "currencyId": 121,
        "orgId": 1000001,
        "name": "DID/DDI"
    }'

    curl -XPOST -H "Content-type: application/json" http://localhost:9090/chump/v2/numbers -d \
    '{
        "requests": [
            {
                "number": "6494297024",
                "domain": "conversant.co.nz",
                "priceListVersionId": 1000000,
                "countryId": "147",
                "countryCode": "64",
                "areaCode": "9429",
                "areaCodeDescription": "Auckland - Red Beach Test",
                "freeMinutes": "10",
                "perMinuteCharge": "0",
                "businessPartnerId": 1000076,
                "setupCost": "1",
                "monthlyCharge": "2",
                "currencyId": 121,
                "orgId": 1000001,
                "name": "DID/DDI"
            },
            {
                "number": "6494297025",
                "domain": "conversant.co.nz",
                "priceListVersionId": 1000000,
                "countryId": "147",
                "countryCode": "64",
                "areaCode": "9429",
                "areaCodeDescription": "Auckland - Red Beach Test",
                "freeMinutes": "10",
                "perMinuteCharge": "0",
                "businessPartnerId": 1000076,
                "setupCost": "1",
                "monthlyCharge": "2",
                "currencyId": 121,
                "orgId": 1000001,
                "name": "DID/DDI"
            }
        ]
    }'

## Provision a number (optional - filter.exclude.did, filter.exclude.call, and filter.exclude.inbound) (see swagger docs)

Using filters:
    filter.exclude.did - This prevents the creation of DID subs by the API call.
    filter.exclude.inbound - This prevents the creation/update of inbound user prefs.
    filter.exclude.call - This prevents the creation of call subs by the API call.

    Filters can be combined with '&' (e.g. ?filter.exclude.did=true&filter.exclude.inbound=true)

    An example of filter use would be in the case of provisioning numbers for migration. When doing this,
    you don't want DID subs to be created or the Inbound user preferences to be updated. As a result
    you would use the following filters:
        ?filter.exclude.did=true&filter.exclude.inbound=true

    curl -XPOST -H "Content-type: application/json" http://localhost:9090/chump/v2/numbers/6494297021/provision -d \
    '{
        "realm": "conversant.co.nz",
        "proxy": "c-vm-02.conversant.co.nz",
        "businessPartnerId": 1000076,
        "businessPartnerLocationId": 1000014,
        "startDate": "2015-02-01",
        "paidUntilDate": "2015-03-01",
        "orgId": 1000001
    }'

    curl -XPOST -H "Content-type: application/json" http://localhost:9090/chump/v2/numbers/provision -d \
    '{
        "numbers": ["6494297022", "6494297023"],
        "realm": "conversant.co.nz",
        "proxy": "c-vm-02.conversant.co.nz",
        "businessPartnerId": 1000076,
        "businessPartnerLocationId": 1000014,
        "startDate": "2015-02-01",
        "paidUntilDate": "2015-03-01",
        "orgId": 1000001
    }'

    curl -XPOST -H "Content-type: application/json" http://localhost:9090/chump/v2/numbers/6494297024/provision -d \
    '{
        "requests": [
            {
                "number": "6494297024",
                "realm": "conversant.co.nz",
                "proxy": "c-vm-02.conversant.co.nz",
                "businessPartnerId": 1000076,
                "businessPartnerLocationId": 1000014,
                "startDate": "2015-02-01",
                "paidUntilDate": "2015-03-01",
                "orgId": 1000001
            },
            {
                "number": "6494297025",
                "realm": "conversant.co.nz",
                "proxy": "c-vm-02.conversant.co.nz",
                "businessPartnerId": 1000076,
                "businessPartnerLocationId": 1000014,
                "startDate": "2015-02-01",
                "paidUntilDate": "2015-03-01",
                "orgId": 1000001
            }
        ]
    }'

## Subscribe a number (optional - filter.exclude.did)
    curl -XPOST -H "Content-type: application/json" http://localhost:9090/chump/v2/numbers/6494297021/subscribe -d \
    '{
        "realm": "conversant.co.nz",
        "proxy": "c-vm-02.conversant.co.nz",
        "businessPartnerId": 1000076,
        "businessPartnerLocationId": 1000014,
        "startDate": "2015-02-01",
        "paidUntilDate": "2015-03-01",
        "orgId": 1000001
    }'

    curl -XPOST -H "Content-type: application/json" http://localhost:9090/chump/v2/numbers/subscribe -d \
    '{
        "numbers": ["6494297022", "6494297023"],
        "realm": "conversant.co.nz",
        "proxy": "c-vm-02.conversant.co.nz",
        "businessPartnerId": 1000076,
        "businessPartnerLocationId": 1000014,
        "startDate": "2015-02-01",
        "paidUntilDate": "2015-03-01",
        "orgId": 1000001
    }'

    curl -XPOST -H "Content-type: application/json" http://localhost:9090/chump/v2/numbers/6494297024/subscribe -d \
    '{
        "requests": [
            {
                "number": "6494297024",
                "realm": "conversant.co.nz",
                "proxy": "c-vm-02.conversant.co.nz",
                "businessPartnerId": 1000076,
                "businessPartnerLocationId": 1000014,
                "startDate": "2015-02-01",
                "paidUntilDate": "2015-03-01",
                "orgId": 1000001
            },
            {
                "number": "6494297025",
                "realm": "conversant.co.nz",
                "proxy": "c-vm-02.conversant.co.nz",
                "businessPartnerId": 1000076,
                "businessPartnerLocationId": 1000014,
                "startDate": "2015-02-01",
                "paidUntilDate": "2015-03-01",
                "orgId": 1000001
            }
        ]
    }'

## Set number as primary caller id  (see swagger docs)
    curl -XPUT -H "Content-type: application/json" http://localhost:9090/chump/v2/numbers/6494297026/callerId -d \
    '{
        "businessPartnerId": 1000076,
        "realm": "conversant.co.nz",
        "startDate": "2015-07-06"
    }'

## Provision an order (optional - filter.exclude.did and filter.exclude.call)  (see swagger docs)
    curl -XPOST -H "Content-type: application/json" http://localhost:9090/chump/v2/orders/80088/provision -d \
    '{
        "realm": "conversant.co.nz",
        "proxy": "c-vm-02.conversant.co.nz"
    }'

# REST - v1
## Get status (see swagger docs)
    curl http://localhost:9090/chump/v1/status

## Create a number (optional - filter.exclude.did) (see swagger docs)
    curl -XPOST -H "Content-type: application/json" http://localhost:9090/chump/v1/numbers -d \
    '{
        "number": "6494297021",
        "domain": "conversant.co.nz",
        "priceListVersionId": 1000000,
        "countryId": "147",
        "countryCode": "64",
        "areaCode": "9429",
        "areaCodeDescription": "Auckland - Red Beach Test",
        "freeMinutes": "10",
        "perMinuteCharge": "0",
        "businessPartnerId": 1000076,
        "setupCost": "1",
        "monthlyCharge": "2",
        "currencyId": 121,
        "orgId": 1000001,
        "name": "DID/DDI"
    }'

    curl -XPOST -H "Content-type: application/json" http://localhost:9090/chump/v1/numbers -d \
    '{
        "numbers": ["6494297022", "6494297023"],
        "domain": "conversant.co.nz",
        "priceListVersionId": 1000000,
        "countryId": "147",
        "countryCode": "64",
        "areaCode": "9429",
        "areaCodeDescription": "Auckland - Red Beach Test",
        "freeMinutes": "10",
        "perMinuteCharge": "0",
        "businessPartnerId": 1000076,
        "setupCost": "1",
        "monthlyCharge": "2",
        "currencyId": 121,
        "orgId": 1000001,
        "name": "DID/DDI"
    }'

    curl -XPOST -H "Content-type: application/json" http://localhost:9090/chump/v1/numbers -d \
    '{
        "requests": [
            {
                "number": "6494297024",
                "domain": "conversant.co.nz",
                "priceListVersionId": 1000000,
                "countryId": "147",
                "countryCode": "64",
                "areaCode": "9429",
                "areaCodeDescription": "Auckland - Red Beach Test",
                "freeMinutes": "10",
                "perMinuteCharge": "0",
                "businessPartnerId": 1000076,
                "setupCost": "1",
                "monthlyCharge": "2",
                "currencyId": 121,
                "orgId": 1000001,
                "name": "DID/DDI"
            },
            {
                "number": "6494297025",
                "domain": "conversant.co.nz",
                "priceListVersionId": 1000000,
                "countryId": "147",
                "countryCode": "64",
                "areaCode": "9429",
                "areaCodeDescription": "Auckland - Red Beach Test",
                "freeMinutes": "10",
                "perMinuteCharge": "0",
                "businessPartnerId": 1000076,
                "setupCost": "1",
                "monthlyCharge": "2",
                "currencyId": 121,
                "orgId": 1000001,
                "name": "DID/DDI"
            }
        ]
    }'

## Provision a number (optional - filter.exclude.did, filter.exclude.call, and filter.exclude.inbound) (see swagger docs)
    curl -XPOST -H "Content-type: application/json" http://localhost:9090/chump/v1/numbers/6494297021/provision -d \
    '{
        "realm": "conversant.co.nz",
        "proxy": "c-vm-02.conversant.co.nz",
        "businessPartnerId": 1000076, 
        "businessPartnerLocationId": 1000014, 
        "startDate": "2015-02-01",
        "paidUntilDate": "2015-03-01",
        "orgId": 1000001
    }'

    curl -XPOST -H "Content-type: application/json" http://localhost:9090/chump/v1/numbers/provision -d \
    '{
        "numbers": ["6494297022", "6494297023"],
        "realm": "conversant.co.nz",
        "proxy": "c-vm-02.conversant.co.nz",
        "businessPartnerId": 1000076,
        "businessPartnerLocationId": 1000014,
        "startDate": "2015-02-01",
        "paidUntilDate": "2015-03-01",
        "orgId": 1000001
    }'

    curl -XPOST -H "Content-type: application/json" http://localhost:9090/chump/v1/numbers/6494297024/provision -d \
    '{
        "requests": [
            {
                "number": "6494297024",
                "realm": "conversant.co.nz",
                "proxy": "c-vm-02.conversant.co.nz",
                "businessPartnerId": 1000076,
                "businessPartnerLocationId": 1000014,
                "startDate": "2015-02-01",
                "paidUntilDate": "2015-03-01",
                "orgId": 1000001
            },
            {
                "number": "6494297025",
                "realm": "conversant.co.nz",
                "proxy": "c-vm-02.conversant.co.nz",
                "businessPartnerId": 1000076,
                "businessPartnerLocationId": 1000014,
                "startDate": "2015-02-01",
                "paidUntilDate": "2015-03-01",
                "orgId": 1000001
            }
        ]
    }'

## Subscribe a number (optional - filter.exclude.did)
    curl -XPOST -H "Content-type: application/json" http://localhost:9090/chump/v1/numbers/6494297021/subscribe -d \
    '{
        "realm": "conversant.co.nz",
        "proxy": "c-vm-02.conversant.co.nz",
        "businessPartnerId": 1000076,
        "businessPartnerLocationId": 1000014,
        "startDate": "2015-02-01",
        "paidUntilDate": "2015-03-01",
        "orgId": 1000001
    }'

    curl -XPOST -H "Content-type: application/json" http://localhost:9090/chump/v1/numbers/subscribe -d \
    '{
        "numbers: ["6494297022", "6494297023"],
        "realm": "conversant.co.nz",
        "proxy": "c-vm-02.conversant.co.nz",
        "businessPartnerId": 1000076,
        "businessPartnerLocationId": 1000014,
        "startDate": "2015-02-01",
        "paidUntilDate": "2015-03-01",
        "orgId": 1000001
    }'

    curl -XPOST -H "Content-type: application/json" http://localhost:9090/chump/v1/numbers/6494297022/subscribe -d \
    '{
        "requests": [
            {
                "number": "6494297024",
                "realm": "conversant.co.nz",
                "proxy": "c-vm-02.conversant.co.nz",
                "businessPartnerId": 1000076,
                "businessPartnerLocationId": 1000014,
                "startDate": "2015-02-01",
                "paidUntilDate": "2015-03-01",
                "orgId": 1000001
            },
            {
                "number": "6494297025",
                "realm": "conversant.co.nz",
                "proxy": "c-vm-02.conversant.co.nz",
                "businessPartnerId": 1000076,
                "businessPartnerLocationId": 1000014,
                "startDate": "2015-02-01",
                "paidUntilDate": "2015-03-01",
                "orgId": 1000001
            }
        ]
    }'

## Set number as primary caller id (see swagger docs)
    curl -XPUT -H "Content-type: application/json" http://localhost:9090/chump/v1/numbers/6494297026/callerId -d \
    '{
        "businessPartnerId": 1000076,
        "startDate": "2015-07-06"
    }'

## Create an order (see swagger docs)
    curl -XPOST -H "Content-type: application/json" http://localhost:9090/chump/v1/orders -d \
    '{
        "businessPartnerId": 1002255,
        "businessPartnerLocationId": 1002380,
        "pricelistId": 1000000,
        "warehouseId": 1000000,
        "datePromised": "2015-07-01",
        "dateOrdered": "2015-06-01",
        "orgId": 1000001
    }'

## Update an order (see swagger docs)
    curl -XPUT -H "Content-type: application/json" http://localhost:9090/chump/v1/orders/52732 -d \
    '{
        "businessPartnerId": 1002255,
        "businessPartnerLocationId": 1002380,
        "pricelistId": 1000000,
        "warehouseId": 1000000,
        "datePromised": "2020-09-01",
        "dateOrdered": "2020-06-01",
        "orgId": 1000001
    }'

## Provision an order (optional - filter.exclude.did and filter.exclude.call) (see swagger docs)
    curl -XPOST -H "Content-type: application/json" http://localhost:9090/chump/v1/orders/52734/provision -d \
    '{
        "realm": "conversant.co.nz",
        "proxy": "c-vm-02.conversant.co.nz"
    }'

## Read an order (see swagger docs)
    curl http://localhost:9090/chump/v1/orders/52732

## Read order lines (optional query parameters - productId and productCategoryId) (see swagger docs)
    curl http://localhost:9090/chump/v1/orders/52732/lines

## Search call records (see swagger docs)
    curl -XPOST -H "Content-type: application/json" http://localhost:9090/chump/v1/callrecords/search -d \
    '{
        "ids": ["22985c454b0977543912753e5d1d3ff0@202.180.76.164", "6cc1347370082741616f98dc286a30e6@conversant.co.nz"]
    }'

## Get a list of orders belonging to a business partner (see swagger docs)
    curl http://localhost:9090/chump/v1/businesspartners/1000009/orders

## Get a list of locations for a business partner
    curl http://localhost:9090/chump/v1/businesspartners/1000009/locations (see swagger docs)


CHANGELOG.md
# Change Log

## 2017-03-17 - 1.20.0
 - Updated WSDLs
 - RAD-2 - Added classified option to BusinessPartnerRadiusAccountsRoute

## 2017-02-16 - 1.19.0
 - Updated WSDLs
 - OPT-20 - Created One Off Payment Route under BP Payments

## 2017-01-26 - 1.18.0
 - Updated WSDLs
 - CHUMP-117 - Fixed updating business partner locations

## 2017-01-25 - 1.17.0
 - Updated WSDLs
 - CHUMP-115 - Added subscription delay

## 2017-01-09 - 1.16.0
 - Updated WSDLs
 - CHUMP-116 - Modified BusinessPartnerLocation to include paymentRule

## 2016-11-28 - 1.15.1
- CHUMP-112 - Fixed deleting bank account by id

## 2016-11-17 - 1.15.0
- CHUMP-111 - Created provision v3 to enable automated number port subscription creation

## 2016-10-07 - 1.14.0
- Updated WSDLs
- CHUMP-110 - Modified delete BP Bank Account to accept Account Name

## 2016-05-12 - 1.13.0
- Updated WSDLs

## 2016-04-21 - 1.12.1
- CHUMP-105 - Fixed UpdateBusinessPartnersRoute to include paymentTermId, priceListId, paymentRule

## 2016-04-14 - 1.12.0
- Updated WSDLs

## 2016-04-09 - 1.11.0
- Updated WSDLs

## 2016-04-07 - 1.10.1
- CHUMP-103 - Fixed createBankAccount to include bankId

## 2016-03-31 - 1.10.0
- CHUMP-95 - Added unsubscribe number(s) route

## 2016-03-17 - 1.9.0
- CHUMP-100 - Added payment term and payment rule to businessPartner

## 2016-03-17 - 1.8.0
- CHUMP-101 - Added priceListId to create and update businessPartner

## 2016-02-25 - 1.7.0
- CHUMP-99 - DeleteBusinessPartnerBankAccountRoute

## 2016-02-22 - 1.6.1
- CHUMP-94 - UserId no longer required for BPBankAccountCreation

## 2016-02-22 - 1.6.0
- CHUMP-98 - Added BusinessPartnerRadiusAccountsSearch

## 2016-02-19 - 1.5.2
- CHUMP-96 - Patched location requests to no longer require cityId and regionId

## 2016-02-19 - 1.5.1
- CHUMP-96 - Added location to BPBankAccount requests

## 2016-02-02 - 1.5.0
- CHUMP-97 - ReadInvoiceTaxLines

## 2016-01-27 - 1.4.0
- CHUMP-96 - ReadBusinessPartnerPaymentsRoute

## 2016-01-26 - 1.3.0
- CHUMP-94 - CreateBPBankAccount

## 2016-01-08 - 1.2.0
- CHUMP-88 - ReadUserBySearchKey

## 2015-12-16 - 1.1.0
- CHUMP-92 - ReadBusinessPartnerSubscriptionsRoute

## 2015-12-06 - 1.0.0
- Initial release
- Swagger documentation
- Adempiere resources
 - Business partners
 - Locations
 - Invoices
 - Orders
 - Numbers
 - Products
 - Subscriptions
 - Users
 - User roles
 - Organisations
- MySQL resources
 - Call records (radacct)
 - Billing customers
 - User preferences


src/main/java/com/conversant/chump/route/v1/OrderRoute.java
package com.conversant.chump.route.v1;

import com.conversant.chump.common.ChumpOperation;
import com.conversant.chump.common.ChumpRoute;
import com.conversant.chump.common.RestOperation;
import com.conversant.chump.model.ApiResponse;
import com.conversant.chump.model.NumberRequest;
import com.conversant.chump.model.ProvisionOrderRequest;
import com.conversant.chump.processor.ApiResponseProcessor;
import com.conversant.chump.processor.StandardResponseRemover;
import com.conversant.chump.route.AdempiereRoute;
import com.conversant.chump.route.v1.numbers.ProvisionNumbersRoute;
import com.conversant.webservice.*;
import org.apache.camel.Exchange;
import org.apache.camel.Processor;
import org.apache.camel.builder.RouteBuilder;
import org.apache.camel.processor.aggregate.AggregationStrategy;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import static com.conversant.chump.common.RestOperation.HttpMethod.GET;
import static com.conversant.chump.common.RestOperation.HttpMethod.POST;
import static com.conversant.chump.util.AdempiereHelper.createLoginRequest;
import static com.conversant.chump.util.Constants.*;

/**
 * Created by jhill on 31/12/14.
 */
@Component
public class OrderRoute implements ChumpRoute {

    /** Base resource */
    private static final String RESOURCE = "/v1/orders";

    private static final String PROVISION_ORDER_CUSTOM = "direct://provisionOrderCustom";
    private static final String PROVISION_NUMBER_PORTS_CUSTOM = "direct://provisionNumberPortsCustom";
    private static final String ORDER = "order";

    /** Provision all numbers and number ports on an order */
    public static final ChumpOperation PROVISION = ChumpOperation.builder()
            .rest(RestOperation.builder()
                    .method(POST)
                    .resource(RESOURCE)
                    .path("/{orderNo}/provision")
                    .requestType(ProvisionOrderRequest.class)
                    .build())
            .trx(false)
            // TODO: Can remove once fix header and path param
            .preProcessors(Arrays.asList(ProvisionOrderRequestProcessor.INSTANCE))
            .to(Arrays.asList(
                    ChumpOperation.pair(ReadOrderRequestProcessor.INSTANCE, AdempiereRoute.READ_ORDER.getUri()),
                    ChumpOperation.pair(ReadOrderDIDsRequestProcessor.INSTANCE, AdempiereRoute.READ_ORDER_DIDS.getUri()),
                    ChumpOperation.pair(ProvisionNumberSplitRequestProcessor.INSTANCE, PROVISION_ORDER_CUSTOM),
                    ChumpOperation.pair(ReadOrderNumberPortsRequestProcessor.INSTANCE, AdempiereRoute.READ_ORDER_NUMBER_PORTS.getUri()),
                    ChumpOperation.pair(OrderNumberPortsSplitRequestProcessor.INSTANCE, PROVISION_NUMBER_PORTS_CUSTOM)))
            .build();

    public static final ChumpOperation LINES = ChumpOperation.builder()
            .rest(RestOperation.builder()
                    .method(GET)
                    .resource(RESOURCE)
                    .path("/{orderNo}/lines")
                    .build())
            .trx(false)
            .to(Arrays.asList(
                    ChumpOperation.pair(ReadOrderRequestHeaderProcessor.INSTANCE, AdempiereRoute.READ_ORDER.getUri()),
                    ChumpOperation.pair(ReadOrderLinesRequestProcessor.INSTANCE, AdempiereRoute.READ_ORDER_LINES.getUri())))
            .postProcessors(Arrays.asList(
                    new StandardResponseRemover("orderLine"), ApiResponseProcessor.INSTANCE))
            .build();

    @Component
    private static final class OrderRouteBuilder extends RouteBuilder {
        @Override
        public void configure() throws Exception {

            // Custom route for fancy split/aggregate logic
            from(PROVISION_ORDER_CUSTOM)

                    // Split on the body, individual ProvisionNumberRequest, and use custom aggregation strategy
                    // which groups individual ApiResponse's into a list
                    .split(body(), ProvisionNumberAggregationStrategy.INSTANCE)

                    // Call provision number for each split request
                    .to(ProvisionNumbersRoute.PROVISION.getUri()).end()

                    // Process final result of custom aggregation strategy into a single ApiResponse
                    .process(AggregatedApiResponseProcessor.INSTANCE);

            // Custom route for fancy split/aggregate logic
            from(PROVISION_NUMBER_PORTS_CUSTOM)

                    // Split on the body, individual CreateNumberPortSubscriptionRequest, and use custom aggregation strategy
                    // which groups individual ApiResponse's into a list
                    .split(body(), ProvisionNumberPortsAggregationStrategy.INSTANCE)

                            // Call provision number for each split request
                    .to(AdempiereRoute.CREATE_NUMBER_PORT_SUBSCRIPTION.getUri()).end()

                    // Process final result of custom aggregation strategy into a single ApiResponse
                    .process(AggregatedApiResponseProcessor.INSTANCE);
        }
    }

    private static final class ProvisionOrderRequestProcessor implements Processor {

        public static final Processor INSTANCE = new ProvisionOrderRequestProcessor();

        @Override
        public void process(Exchange exchange) throws Exception {

            // TODO: How get into request POJO automatically? Generic processor?
            ProvisionOrderRequest request = exchange.getIn().getBody(ProvisionOrderRequest.class);
            if (request.getOrderNo() == null)
                request.setOrderNo((String) exchange.getIn().getHeader("orderNo"));
        }
    }

    private static final class ReadOrderRequestProcessor implements Processor {

        public static final Processor INSTANCE = new ReadOrderRequestProcessor();

        @Override
        public void process(Exchange exchange) throws Exception {

            ProvisionOrderRequest request = exchange.getProperty(ProvisionOrderRequest.class.getName(), ProvisionOrderRequest.class);

            ReadOrderRequest readOrderRequest = new ReadOrderRequest();
            readOrderRequest.setLoginRequest(createLoginRequest(exchange, TYPE_READ_ORDER));
            readOrderRequest.setDocumentNo(request.getOrderNo());

            exchange.getIn().setBody(readOrderRequest);
        }
    }

    private static final class ReadOrderDIDsRequestProcessor implements Processor {

        public static final Processor INSTANCE = new ReadOrderDIDsRequestProcessor();

        @Override
        public void process(Exchange exchange) throws Exception {

            ReadOrderResponse response = exchange.getIn().getBody(ReadOrderResponse.class);

            ReadOrderDIDsRequest readOrderDIDsRequest = new ReadOrderDIDsRequest();
            readOrderDIDsRequest.setLoginRequest(createLoginRequest(exchange, TYPE_READ_ORDER_DIDS));
            readOrderDIDsRequest.setOrderId(response.getOrder().getOrderId());

            exchange.getIn().setBody(readOrderDIDsRequest);

            exchange.setProperty(ORDER, response.getOrder());
        }
    }

    private static final class ProvisionNumberSplitRequestProcessor implements Processor {

        public static final Processor INSTANCE = new ProvisionNumberSplitRequestProcessor();

        @Override
        public void process(Exchange exchange) throws Exception {

            ProvisionOrderRequest provisionOrderRequest = exchange.getProperty(ProvisionOrderRequest.class.getName(), ProvisionOrderRequest.class);
            Order order = exchange.getProperty(ORDER, Order.class);

            List<NumberRequest> requests = exchange.getIn().getBody(ReadOrderDIDsResponse.class).getDids().stream()
                    .map(did -> {

                        NumberRequest request = new NumberRequest();
                        request.setNumber(did);
                        request.setRealm(provisionOrderRequest.getRealm());
                        request.setProxy(provisionOrderRequest.getProxy());
                        request.setBusinessPartnerId(order.getBusinessPartnerId());
                        request.setBusinessPartnerLocationId(order.getBusinessPartnerLocationId());
                        request.setStartDate(order.getDatePromised());
                        request.setPaidUntilDate(order.getDatePromised());
                        request.setOrgId(order.getOrgId());

                        return request;
                    })
                    .collect(Collectors.toList());

            // Set body as list of requests for use with split by body
            exchange.getIn().setBody(requests);
        }
    }

    private static final class AggregatedApiResponseProcessor implements Processor {

        public static final Processor INSTANCE = new AggregatedApiResponseProcessor();

        @Override
        public void process(Exchange exchange) throws Exception {

            // Check if existing ApiResponse set
            ApiResponse existing = exchange.getProperty(PROPERTY_API_RESPONSE, ApiResponse.class);

            // Check if any ApiResponse's failed
            boolean failure = exchange.getIn().getBody(List.class).stream().anyMatch(o -> o instanceof ApiResponse && ((ApiResponse) o).getCode() != 200);

            // If found existing and it contains a failure then fail this response
            if (existing != null && existing.getCode() != ApiResponse.SUCCESS) {
                failure = true;
            }

            // Create ApiResponse and all list of responses
            ApiResponse response = failure ? ApiResponse.badRequest() : ApiResponse.success();
            response.setResponses(exchange.getIn().getBody(List.class));

            // If found existing then add responses
            if (existing != null && existing.getResponses() != null) {
                response.getResponses().addAll(existing.getResponses());
            }

            // Set as exchange property to be picked up by ApiResponseProcessor
            exchange.setProperty(PROPERTY_API_RESPONSE, response);
        }
    }

    private static final class ProvisionNumberAggregationStrategy implements AggregationStrategy {

        public static final AggregationStrategy INSTANCE = new ProvisionNumberAggregationStrategy();

        @Override
        public Exchange aggregate(Exchange oldExchange, Exchange newExchange) {

            ApiResponse response = newExchange.getIn().getBody(ApiResponse.class);

            if (response.getCode() == ApiResponse.SUCCESS) {
                NumberRequest request = newExchange.getProperty(NumberRequest.class.getName(), NumberRequest.class);
                response.setMessage("Provisioned " + request.getNumber());
            }

            // First invocation
            if (oldExchange == null) {

                List<ApiResponse> responses = new ArrayList<>();
                responses.add(response);

                newExchange.getIn().setBody(responses);

                return newExchange;
            }

            oldExchange.getIn().getBody(List.class).add(response);

            return oldExchange;
        }
    }

    private static final class ReadOrderNumberPortsRequestProcessor implements Processor {

        public static final Processor INSTANCE = new ReadOrderNumberPortsRequestProcessor();

        @Override
        public void process(Exchange exchange) throws Exception {

            Order order = exchange.getProperty(ORDER, Order.class);

            ReadOrderNumberPortsRequest readOrderNumberPortsRequest = new ReadOrderNumberPortsRequest();
            readOrderNumberPortsRequest.setLoginRequest(createLoginRequest(exchange, TYPE_READ_ORDER_NUMBER_PORTS));
            readOrderNumberPortsRequest.setOrderId(order.getOrderId());

            exchange.getIn().setBody(readOrderNumberPortsRequest);
        }
    }

    private static final class OrderNumberPortsSplitRequestProcessor implements Processor {

        public static final Processor INSTANCE = new OrderNumberPortsSplitRequestProcessor();

        @Override
        public void process(Exchange exchange) throws Exception {

            Order order = exchange.getProperty(ORDER, Order.class);

            List<CreateNumberPortSubscriptionRequest> requests = exchange.getIn().getBody(ReadOrderNumberPortsResponse.class).getNumbers().stream()
                    .map(did -> {

                        CreateNumberPortSubscriptionRequest request = new CreateNumberPortSubscriptionRequest();
                        request.setLoginRequest(createLoginRequest(exchange, TYPE_CREATE_NUMBER_PORT_SUBSCRIPTION));
                        request.setNumber(did);
                        request.setBusinessPartnerId(order.getBusinessPartnerId());
                        request.setBusinessPartnerLocationId(order.getBusinessPartnerLocationId());
                        request.setOrgId(order.getOrgId());

                        return request;
                    })
                    .collect(Collectors.toList());

            // Set body as list of requests for use with split by body
            exchange.getIn().setBody(requests);
        }
    }

    private static final class ProvisionNumberPortsAggregationStrategy implements AggregationStrategy {

        public static final AggregationStrategy INSTANCE = new ProvisionNumberPortsAggregationStrategy();

        @Override
        public Exchange aggregate(Exchange oldExchange, Exchange newExchange) {

            StandardResponse response = newExchange.getIn().getBody(StandardResponse.class);
            CreateNumberPortSubscriptionRequest request = newExchange.getProperty(CreateNumberPortSubscriptionRequest.class.getName(), CreateNumberPortSubscriptionRequest.class);

            ApiResponse apiResponse;
            if (response.isSuccess()) {
                apiResponse = ApiResponse.success();
                apiResponse.setMessage("Created number port subscription for " + request.getNumber());
            }
            else {
                apiResponse = ApiResponse.error();
                apiResponse.setMessage("Failed to create number port subscription for " + request.getNumber());
            }

            // First invocation
            if (oldExchange == null) {

                List<ApiResponse> responses = new ArrayList<>();
                responses.add(apiResponse);

                newExchange.getIn().setBody(responses);

                return newExchange;
            }

            oldExchange.getIn().getBody(List.class).add(apiResponse);

            return oldExchange;
        }
    }

    private static final class ReadOrderRequestHeaderProcessor implements Processor {

        public static final Processor INSTANCE = new ReadOrderRequestHeaderProcessor();

        @Override
        public void process(Exchange exchange) throws Exception {

            ReadOrderRequest readOrderRequest = new ReadOrderRequest();
            readOrderRequest.setLoginRequest(createLoginRequest(exchange, TYPE_READ_ORDER, ADEMPIERE_USER_INTALIO));
            readOrderRequest.setDocumentNo((String) exchange.getIn().getHeader("orderNo"));

            if (exchange.getIn().getHeader("productId") != null) {
                exchange.setProperty("productId", Integer.parseInt((String) exchange.getIn().getHeader("productId")));
            }

            if (exchange.getIn().getHeader("productCategoryId") != null) {
                exchange.setProperty("productCategoryId", Integer.parseInt((String) exchange.getIn().getHeader("productCategoryId")));
            }

            exchange.getIn().setBody(readOrderRequest);
        }
    }

    private static class ReadOrderLinesRequestProcessor implements Processor {

        public static final Processor INSTANCE = new ReadOrderLinesRequestProcessor();

        @Override
        public void process(Exchange exchange) throws Exception {

            ReadOrderResponse response = exchange.getIn().getBody(ReadOrderResponse.class);

            ReadOrderLinesRequest request = new ReadOrderLinesRequest();
            request.setLoginRequest(createLoginRequest(exchange, TYPE_READ_ORDER_LINES, ADEMPIERE_USER_INTALIO));
            request.setOrderId(response.getOrder().getOrderId());

            Integer productId = exchange.getProperty("productId", Integer.class);
            if (productId != null) {
                request.setProductId(productId);
            }

            Integer productCategoryId = exchange.getProperty("productCategoryId", Integer.class);
            if (productCategoryId != null) {
                request.setProductCategoryId(productCategoryId);
            }

            exchange.getIn().setBody(request);
        }
    }
}
